node/network/bitfield-distribution/src/lib.rs

```diff
@@ -0,0 +1,684 @@
+			for input in input.into_iter() {
+				handle.send(input.into());
+			}
+
+			// @todo cannot clone or move `ctx`
+			//
+			// let completion = BitfieldDistribution::start(BitfieldDistribution, ctx)
+			//     .future
+			//     .timeout(Duration::from_millis(1000))
+			//     .await;
+
+			// while let Ok(rxd) = ctx.recv().await {
+			//     // @todo impl expectation checks against a hashmap
+			//     dbg!(rxd);
+			// }
+		});
+	}
+
+	/// A very limited tracker, only interested in the relay parent of the
+	/// given message, which must be signed by `validator` and a set of peers
+	/// which are also only interested in that relay parent.
+	fn prewarmed_tracker(
+		validator: ValidatorId,
+		signing_context: SigningContext,
+		known_message: BitfieldGossipMessage,
+		peers: Vec<PeerId>,
+	) -> Tracker {
+		let relay_parent = known_message.relay_parent.clone();
+		Tracker {
+			per_relay_parent: hashmap! {
+				relay_parent.clone() =>
+					PerRelayParentData {
+						signing_context,
+						validator_set: vec![validator.clone()],
+						one_per_validator: hashmap! {
+							validator.clone() => known_message.clone(),
+						},
+						message_sent_to_peer: hashmap! {},
+					},
+			},
+			peer_views: peers
+				.into_iter()
+				.map(|peer| (peer, view!(relay_parent)))
+				.collect(),
+			view: view!(relay_parent),
+		}
+	}
+
+	#[test]
+	fn receive_invalid_signature() {
+		let _ = env_logger::builder()
+			.filter(None, log::LevelFilter::Trace)
+			.is_test(true)
+			.try_init();
+
+		let hash_a: Hash = [0; 32].into();
+		let hash_b: Hash = [1; 32].into(); // other
+
+		let peer_a = PeerId::random();
+		let peer_b = PeerId::random();
+		assert_ne!(peer_a, peer_b);
+
+		let signing_context = SigningContext {
+			session_index: 1,
+			parent_hash: hash_a.clone(),
+		};
+
+		// validator 0 key pair
+		let (validator_pair, _seed) = ValidatorPair::generate();
+		let validator = validator_pair.public();
+
+		// another validator not part of the validatorset
+		let (mallicious, _seed) = ValidatorPair::generate();
+
+		let payload = AvailabilityBitfield(bitvec![bitvec::order::Lsb0, u8; 1u8; 32]);
+		let signed =
+			Signed::<AvailabilityBitfield>::sign(payload, &signing_context, 0, &mallicious);
+
+		let msg = BitfieldGossipMessage {
+			relay_parent: hash_a.clone(),
+			signed_availability: signed.clone(),
+		};
+
+		let pool = sp_core::testing::SpawnBlockingExecutor::new();
+		let (mut ctx, mut handle) =
+			subsystem_test::make_subsystem_context::<BitfieldDistributionMessage, _>(pool);
+
+		let mut tracker = prewarmed_tracker(
+			validator.clone(),
+			signing_context.clone(),
+			msg.clone(),
+			vec![peer_b.clone()],
+		);
+
+		executor::block_on(async move {
+			let _res = handle_network_msg(&mut ctx, &mut tracker, NetworkBridgeEvent::PeerMessage(peer_b.clone(), msg.encode()))
+				.timeout(Duration::from_millis(10))
+				.await
+				.expect("10ms is more than enough for sending messages. qed")
+				.expect("There are no error values. qed");
```

Comment:
panicking with `unwrap` or without `qed` in general is fine in tests, btw :)

Reply:
Will skip the `qed` part in the future, I just prefer verbose `panic`s in test cases :)